### ISSUE TITLE
vw compilation warnings

### ISSUE THREAD

**Issue by arielf** [VowpalWabbit/vowpal_wabbit#1630]
May be related to https://github.com/VowpalWabbit/vowpal_wabbit/issues/1629

When building `vw` (with clang pedantic checks) I get these warnings:
```
clang++ -static -std=c++11   -D__extern_always_inline=inline -Wall -pedantic -DNDEBUG -O3 -fomit-frame-pointer -fno-strict-aliasing -msse2 -mfpmath=sse  -D_FILE_OFFSET_BITS=64 -I /usr/local/include/boost -I /usr/include -I ../rapidjson/include -fPIC  -I ../rapidjson/include -I ../explore -c csoaa.cc -o csoaa.o
In file included from ect.cc:17:
In file included from ./reductions.h:14:
In file included from ./global_data.h:32:
./parser_helper.h:73:51: warning: lambda capture 'def' is not used [-Wunused-lambda-capture]
                        ->notifier([this, option, def] (T arg)
                                                  ^
ect.cc:361:8: note: in instantiation of function template specialization 'arguments::keep<unsigned long>' requested here
      .keep("error", data->errors, (uint64_t)0, "errors allowed by ECT").missing())
       ^
1 warning generated.


# And later:

clang++ -static -std=c++11   -D__extern_always_inline=inline -Wall -pedantic -DNDEBUG -O3 -fomit-frame-pointer -fno-strict-aliasing -msse2 -mfpmath=sse  -D_FILE_OFFSET_BITS=64 -I /usr/local/include/boost -I /usr/include -I ../rapidjson/include -fPIC  -I ../rapidjson/include -I ../explore -c parse_args.cc -o parse_args.o
In file included from search_dep_parser.cc:6:
In file included from ./search_dep_parser.h:7:
In file included from ./search.h:7:
In file included from ./global_data.h:32:
./parser_helper.h:73:51: warning: lambda capture 'def' is not used [-Wunused-lambda-capture]
                        ->notifier([this, option, def] (T arg)
                                                  ^
search_dep_parser.cc:56:6: note: in instantiation of function template specialization 'arguments::keep<unsigned long>' requested here
    .keep("root_label", data->root_label, (size_t)8, "Ensure that there is only one root in each sentence")
     ^
In file included from search_dep_parser.cc:6:
In file included from ./search_dep_parser.h:7:
In file included from ./search.h:7:
In file included from ./global_data.h:32:
./parser_helper.h:73:51: warning: lambda capture 'def' is not used [-Wunused-lambda-capture]
                        ->notifier([this, option, def] (T arg)
                                                  ^
search_dep_parser.cc:57:6: note: in instantiation of function template specialization 'arguments::keep<unsigned int>' requested here
    .keep("num_label", data->num_label, (uint32_t)12, "Number of arc labels")
     ^
2 warnings generated.
```

Similar warnings keep repeating many times.

**Comment by jackgerrits** [comment]
There are quite a few warnings present at the moment, especially since I raised the warning level in the change to CMake. These warnings will be fixed as part of a refactor of options I am working on

**Comment by JohnLangford** [comment]
We should definitely clean these up.   Warnings tend to end up hiding errors ...

**Comment by jackgerrits** [comment]
These are resolved with #1710 and #1713
